feature/savings/src/commonMain/kotlin/com/mifos/feature/savings/savingsAccountv2/SavingsAccountScreen.kt

```diff
@@ -152,6 +178,156 @@ private fun NewSavingsAccountDialog(
                 onclick = { onAction(SavingsAccountAction.Retry) },
             )
         }
+
+        is SavingsAccountState.DialogState.AddNewCharge -> AddNewChargeDialog(
+            isEdit = state.dialogState.edit,
+            state = state,
+            onAction = onAction,
+            index = state.dialogState.index,
+        )
+
+        is SavingsAccountState.DialogState.ShowCharges -> ShowChargesDialog(
+            state = state,
+            onAction = onAction,
+        )
+
         null -> Unit
     }
 }
+
+@OptIn(ExperimentalTime::class, ExperimentalMaterial3Api::class)
+@Composable
+private fun AddNewChargeDialog(
+    isEdit: Boolean,
+    index: Int = -1,
+    state: SavingsAccountState,
+    onAction: (SavingsAccountAction) -> Unit,
+) {
+    LaunchedEffect(state.chargeAmount) {
+        val amountError = doubleNumberValidator(state.chargeAmount)
+        onAction(SavingsAccountAction.OnChargesAmountChangeError(amountError))
+    }
+    fun isSelectableDate(utcTimeMillis: Long): Boolean {
+        return utcTimeMillis >= Clock.System.now().toEpochMilliseconds().minus(86_400_000L)
+    }
+    AddChargeBottomSheet(
+        title = if (isEdit) {
+            stringResource(Res.string.step_charges_edit_charge)
+        } else {
+            stringResource(Res.string.step_charges_add_new) + " " + stringResource(Res.string.step_charges)
+        },
+        confirmText = if (isEdit) {
+            stringResource(Res.string.step_charges_edit_charge)
+        } else {
+            stringResource(Res.string.step_charges_add)
+        },
+        dismissText = stringResource(Res.string.feature_savings_cancel),
+        showDatePicker = state.showChargesDatePick,
+        selectedChargeName = if (state.chooseChargeIndex == -1) {
+            ""
+        } else {
+            state.savingsProductTemplate?.chargeOptions?.get(state.chooseChargeIndex)?.name ?: ""
```

Comment:
List access occurs without verifying index bounds, which may cause crashes.

Suggested fix: Use safe access.
```suggestion
           state.savingsProductTemplate?.chargeOptions?.getOrNull(state.chooseChargeIndex)?.name ?: ""
```